File: solutions/211.design-add-and-search-words-data-structure.java
class WordDictionary {
    WordDictionary[] kids = new WordDictionary[26];
    boolean isEnd;
    public WordDictionary() {
        
    }
    
    public void addWord(String word) {
        var cur = this;
        for(var c : word.toCharArray()) {
            int index = c-'a';
            if(cur.kids[index] == null) cur.kids[index] = new WordDictionary();
            cur = cur.kids[index];
        }
        cur.isEnd = true;
    }
    
    public boolean search(String word) {
        var cur = this;
        for(int i=0; i<word.length(); ++i) {
            char c = word.charAt(i);
            if(c=='.') {
                for(var kid : cur.kids) { // watchout: cur.kids not this.kids
                    if(kid!=null && kid.search(word.substring(i+1, word.length()))) {
                        return true;
                    }
                }
                return false;
            } else {
                int index = c-'a';
                if(cur.kids[index] == null) return false;
                cur = cur.kids[index];
            }
        }
        return cur.isEnd;
    }
    
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: solutions/211.design-add-and-search-words-data-structure.py
class WordDictionary:

    def __init__(self):
        self.hash = {}
        self.is_end = False

    def __getitem__(self, key):
        return self.hash[key]

    def __setitem__(self, key, value) -> None:
        self.hash[key] = value

    def addWord(self, word: str) -> None:
        node = self
        for c in word:
            node[c] = node[c] if c in node.hash else WordDictionary()
            node = node[c]
        node.is_end = True


    def search(self, word: str, start=0) -> bool:
        node = self
        for i in range(start, len(word)):
            c = word[i]
            if c == '.':
                for key, kid_node in node.hash.items():
                    if kid_node.search(word, i+1): return True
                return False

            if c not in node.hash: return False
            node = node[c]
        return node.is_end



# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
